recipes/experimental/long-context/H2O/README.md

```diff
@@ -0,0 +1,90 @@
+
+```
+python run_summarization.py \
+--input-path data/summarization/xsum.jsonl \
+--output-path summarization_output/xsum_h2o.jsonl \
+--model-name meta-llama/Llama-2-7b-hf \
+--enable_h2o_generation 
+```
+
+##### **Results**
+
+Expected results on XSUM (Rouge-2 score) from the above scripts on Llama-2/3 models. The sequence length of inputs are ~2k, thus KV cache size larger than 2048 represents the full cache performance.
+
+| KV Cache Size | 64     | 128    | 256    | 512    | 1024   | 2048   | 4096   | 8192   |
+| ------------- | ------ | ------ | ------ | ------ | ------ | ------ | ------ | ------ |
+| Llama-2-7B    | 0.0439 | 0.1127 | 0.1148 | 0.1182 | 0.1170 | 0.1164 | 0.1164 | 0.1164 |
+| Llama-2-13B   | 0.1180 | 0.1217 | 0.1243 | 0.1291 | 0.1302 | 0.1332 | 0.1332 | 0.1332 |
+| Llama-3-8B    | 0.1107 | 0.1189 | 0.1200 | 0.1347 | 0.1290 | 0.1311 | 0.1311 | 0.1311 |
+
+### Evaluation on "Needle in a Haystack" Analysis
+
+The following example runs inference of Llama-3-8b-instruct on "Needle in a haystack" test. The test is modified from [https://github.com/gkamradt/LLMTest_NeedleInAHaystack](). Please follow the original repository for installing necessary packages. We're using `--enable_h2o_generation` to enable H2O algorithm that only keeps heavy-hitter and the local KV pairs. Use `--num_heavy_hitter_tokens` to decide the number of heavy-hitter KV pairs and `--num_window_length `for the KV cache size. The number of local KV pairs equals num_window_length - num_heavy_hitter_tokens. Also, use --enable_position_rolling to enable position rolling in the KV cache size that assign the positions in the KV cache instead of the ones in original sequences. Enabling postional rolling is important when sequence length exceeds the pretrained context windows, e.g., 4K in Llama-2.
+
+```
+# step 1: obtain prompts for evaluation
+python utils/needle_test/prompt.py --model_name meta-llama/Meta-Llama-3-8B-Instruct
+
+
+# step 2: generation predictions of each prompt
+# full model
+python run_needle_haystack_test.py \
+--input-path data/needle_test/Huggingface \
+--output-path needle_test_results/huggingface/llama-3-8b-instruct/ \
+--model-name meta-llama/Meta-Llama-3-8B-Instruct 
+
+# h2o with 2048 kv cache
+python run_needle_haystack_test.py \
+--input-path data/needle_test/Huggingface \
+--output-path needle_test_results/huggingface/llama-3-8b-instruct-h2o-4096/ \
+--model-name meta-llama/Meta-Llama-3-8B-Instruct \
+--enable_h2o_generation \
+--num_window_length 4096 \
+--num_heavy_hitter_tokens 2048
+
+
+# step 3: scoring with gpt4
+export OPENAI_API_KEY=YOUR_API_KEY
+python utils/needle_test/eval.py \
+--input-path needle_test_results/huggingface/llama-3-8b-instruct-h2o-4096\ #path for the prediction results
+--output-path needle_test_results/huggingface/llama-3-8b-instruct-h2o-4096_eval
+
+
+# step 4: visualization
+python utils/needle_test/vis.py \
+--input-path needle_test_results/huggingface/llama-3-8b-instruct-h2o-4096_eval
+```
+
```

Comment:
Can we please add some explanation to clarify specially on the tested length 1M.